File: backend/app/schemas/questions.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


class ContentSegment(BaseModel):
    type: Literal["text", "box", "image"]
    md: Optional[str] = None
    url: Optional[str] = None
    ref: Optional[str] = None


class QuestionCard(BaseModel):
    question_id: int
    meta_short: str
    meta_full: str
    school: Optional[str] = None
    chapter: Optional[str] = None
    difficulty: Optional[str] = None
    points: Optional[float] = None
    is_subjective: bool = False
    error_note: Optional[str] = None
    content_segments: list[ContentSegment]
    choices_text: str = ""
    answer: Optional[str] = None
    answer_display: Optional[str] = None
    solution_segments: list[ContentSegment] = []


class SearchRequest(BaseModel):
    quick_search: str = ""
    regions: list[str] = []
    schools: list[str] = []
    grades: list[int] = []
    years: list[int] = []
    semesters: list[int] = []
    exam_types: list[str] = []
    subjects: list[str] = []
    majors: list[str] = []
    minors: list[str] = []
    difficulties: list[str] = []
    question_type: Literal["all", "choice", "subjective"] = "all"
    keyword: str = ""
    page: int = 0
    page_size: int = 15


class SearchResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[QuestionCard]


class ByIdsRequest(BaseModel):
    question_ids: list[int]


class ByIdsResponse(BaseModel):
    items: list[QuestionCard]


class SearchIdsResponse(BaseModel):
    question_ids: list[int]


class MiniTestRequest(SearchRequest):
    # main.py:931-937 슬라이더 범위(문항 수 6~8, 쉬운문제 비중 0~100 10% 단위) 그대로.
    mini_count: int = Field(default=7, ge=6, le=8)
    mini_easy_pct: int = Field(default=40, ge=0, le=100)


class MiniTestResponse(BaseModel):
    question_ids: list[int]
    pool_size: int

File: frontend/src/api/questions.ts
import { apiFetch } from "./client";
import type { ContentSegment } from "../components/content/QuestionContent";

export interface QuestionCard {
  question_id: number;
  meta_short: string;
  meta_full: string;
  school: string | null;
  chapter: string | null;
  difficulty: string | null;
  points: number | null;
  is_subjective: boolean;
  error_note: string | null;
  content_segments: ContentSegment[];
  choices_text: string;
  answer: string | null;
  answer_display: string | null;
  solution_segments: ContentSegment[];
}

export interface FiltersResponse {
  schools: string[];
  regions: string[];
  years: number[];
  grades: number[];
  semesters: number[];
  exam_types: string[];
  difficulties: string[];
  curriculum: Record<string, Record<string, string[]>>;
}

export type QuestionType = "all" | "choice" | "subjective";

export interface SearchRequest {
  quick_search: string;
  regions: string[];
  schools: string[];
  grades: number[];
  years: number[];
  semesters: number[];
  exam_types: string[];
  subjects: string[];
  majors: string[];
  minors: string[];
  difficulties: string[];
  question_type: QuestionType;
  keyword: string;
  page: number;
  page_size: number;
}

export interface SearchResponse {
  total: number;
  page: number;
  page_size: number;
  items: QuestionCard[];
}

export interface ByIdsResponse {
  items: QuestionCard[];
}

export interface SearchIdsResponse {
  question_ids: number[];
}

export interface MiniTestRequest extends SearchRequest {
  mini_count: number;
  mini_easy_pct: number;
}

export interface MiniTestResponse {
  question_ids: number[];
  pool_size: number;
}

export function fetchFilters(): Promise<FiltersResponse> {
  return apiFetch<FiltersResponse>("/api/filters");
}

export function searchQuestions(req: SearchRequest): Promise<SearchResponse> {
  return apiFetch<SearchResponse>("/api/questions/search", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

export function searchQuestionIds(req: SearchRequest): Promise<SearchIdsResponse> {
  return apiFetch<SearchIdsResponse>("/api/questions/search-ids", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

export function fetchQuestionsByIds(
  questionIds: number[],
  preserveOrder = false,
): Promise<ByIdsResponse> {
  return apiFetch<ByIdsResponse>("/api/questions/by-ids", {
    method: "POST",
    body: JSON.stringify({ question_ids: questionIds, preserve_order: preserveOrder }),
  });
}

export function generateMiniTest(req: MiniTestRequest): Promise<MiniTestResponse> {
  return apiFetch<MiniTestResponse>("/api/questions/mini-test", {
    method: "POST",
    body: JSON.stringify(req),
  });
}
